## fine tuning YOLOv8 model to better detect bounding boxes 

In [52]:
import os
import cv2
import numpy as np
from sklearn.cluster import KMeans
from shutil import copyfile
import json

In [53]:
# def crop_images_from_top_left(source_folder, output_folder, crop_width=1782, crop_height=2455):
#     # Create output folder if it doesn't exist
#     os.makedirs(output_folder, exist_ok=True)

#     # Supported image extensions
#     supported_ext = ('.jpg', '.jpeg', '.png')

#     for filename in os.listdir(source_folder):
#         if filename.lower().endswith(supported_ext):
#             input_path = os.path.join(source_folder, filename)
#             image = cv2.imread(input_path)

#             if image is None:
#                 print(f"Could not read image: {input_path}")
#                 continue

#             # Check if the image is large enough
#             img_height, img_width = image.shape[:2]
#             if img_width < crop_width or img_height < crop_height:
#                 print(f"Image too small to crop: {filename}")
#                 continue

#             # Crop from top-left
#             cropped_image = image[0:crop_height, 0:crop_width]

#             # Create new filename
#             base, ext = os.path.splitext(filename)
#             new_filename = f"{base}_1{ext}"
#             output_path = os.path.join(output_folder, new_filename)

#             # Save cropped image
#             cv2.imwrite(output_path, cropped_image)
#             print(f"Cropped and saved: {output_path}")

# crop_images_from_top_left("new_temp", "new_temp1", crop_width=1782, crop_height=2455)


### Classifying Medical_form front and back page

In [54]:
# import os
# import cv2
# import numpy as np
# from shutil import copyfile
# from sklearn.cluster import KMeans

# # === CONFIGURATION ===
# source_folder = "../data/raw_data/Medical_Care_Form"
# output_folder_0 = "new_data/clust0"
# output_folder_1 = "new_data/clust1"
# image_size = (128, 128)  # Force same size for consistency

# # === PREPARE FOLDERS ===
# os.makedirs(output_folder_0, exist_ok=True)
# os.makedirs(output_folder_1, exist_ok=True)

# # === LOAD AND FLATTEN IMAGES ===
# image_paths = []
# features = []

# for filename in os.listdir(source_folder):
#     if filename.lower().endswith((".jpg", ".jpeg", ".png")):
#         path = os.path.join(source_folder, filename)
#         img = cv2.imread(path, cv2.IMREAD_GRAYSCALE)
#         if img is None:
#             print(f"Skipping unreadable image: {filename}")
#             continue
#         img = cv2.resize(img, image_size)  # Ensures fixed shape
#         img_flat = img.flatten()
#         features.append(img_flat)
#         image_paths.append(path)
#         print(f"Loaded: {filename}")

# print("------------------------------------------------end loading---------------------------------------------")
# features = np.array(features)

# # === APPLY K-MEANS ===
# kmeans = KMeans(n_clusters=2, random_state=42)
# labels = kmeans.fit_predict(features)

# # === SAVE IMAGES INTO CLUSTERS ===
# for path, label in zip(image_paths, labels):
#     filename = os.path.basename(path)
#     dest_folder = output_folder_0 if label == 0 else output_folder_1
#     copyfile(path, os.path.join(dest_folder, filename))
#     print(f"Image saved to cluster {label}: {filename}")

# print("✅ Images clustered and saved into folders!")


## first we will annotate images using roboflow then convert json annotation to yolo format
## Create a label for each image 

In [66]:
# old_temp1
# json_data = {
#     "boxes": [
#         {
#             "id": "1",
#             "label": "nom et prenom de adherent",
#             "x": "977.62",
#             "y": "323.59",
#             "width": "838.33",
#             "height": "115.50",
#             "confidence": None
#         },
#         {
#             "id": "2",
#             "label": "numero cin ou passeport",
#             "x": "823.95",
#             "y": "401.83",
#             "width": "577.15",
#             "height": "91.98",
#             "confidence": None
#         },
#         {
#             "id": "3",
#             "label": "addresse de ladherent",
#             "x": "954.31",
#             "y": "471.77",
#             "width": "1036.32",
#             "height": "107.24",
#             "confidence": None
#         },
#         {
#             "id": "4",
#             "label": "addresse de ladherent",
#             "x": "910.44",
#             "y": "548.39",
#             "width": "1548.58",
#             "height": "112.80",
#             "confidence": None
#         },
#         {
#             "id": "5",
#             "label": "matricule de adherent",
#             "x": "952.23",
#             "y": "625.00",
#             "width": "962.94",
#             "height": "103.21",
#             "confidence": None
#         },
#         {
#             "id": "6",
#             "label": "nom et prenom du malade",
#             "x": "996.54",
#             "y": "1182.09",
#             "width": "756.18",
#             "height": "110.99",
#             "confidence": None
#         },
#         {
#             "id": "7",
#             "label": "date de naissance",
#             "x": "972.54",
#             "y": "1255.02",
#             "width": "985.09",
#             "height": "109.16",
#             "confidence": None
#         },
#         {
#             "id": "8",
#             "label": "date",
#             "x": "256.10",
#             "y": "1637.24",
#             "width": "387.58",
#             "height": "453.59",
#             "confidence": None
#         },
#         {
#             "id": "9",
#             "label": "designation",
#             "x": "508.10",
#             "y": "1639.54",
#             "width": "326.86",
#             "height": "448.97",
#             "confidence": None
#         },
#         {
#             "id": "A",
#             "label": "honoraire",
#             "x": "813.61",
#             "y": "1650.39",
#             "width": "325.06",
#             "height": "475.29",
#             "confidence": None
#         },
#         {
#             "id": "B",
#             "label": "honoraire",
#             "x": "439.07",
#             "y": "2192.90",
#             "width": "619.68",
#             "height": "351.09",
#             "confidence": None
#         }
#     ],
#     "height": 2455,
#     "key": "0756--5637850--20230705_page_0_1.jpg",
#     "width": 1782
# }
# new_temp1
json_data = {
    "boxes": [
        {
            "id": "1",
            "label": "nom et prenom de adherent",
            "x": "986.94",
            "y": "304.05",
            "width": "877.54",
            "height": "102.20",
            "confidence": None
        },
        {
            "id": "2",
            "label": "numero cin ou passeport",
            "x": "834.54",
            "y": "375.03",
            "width": "638.04",
            "height": "89.51",
            "confidence": None
        },
        {
            "id": "3",
            "label": "addresse de ladherent",
            "x": "969.93",
            "y": "451.17",
            "width": "1046.27",
            "height": "94.00",
            "confidence": None
        },
        {
            "id": "4",
            "label": "matricule cnam",
            "x": "765.47",
            "y": "511.84",
            "width": "699.22",
            "height": "88.17",
            "confidence": None
        },
        {
            "id": "5",
            "label": "matricule de adherent",
            "x": "987.97",
            "y": "590.20",
            "width": "1051.43",
            "height": "82.30",
            "confidence": None
        },
        {
            "id": "6",
            "label": "nom et prenom du malade",
            "x": "978.35",
            "y": "1185.27",
            "width": "935.96",
            "height": "108.40",
            "confidence": None
        },
        {
            "id": "7",
            "label": "date de naissance",
            "x": "955.32",
            "y": "1268.09",
            "width": "1102.98",
            "height": "90.25",
            "confidence": None
        },
        {
            "id": "8",
            "label": "date",
            "x": "237.11",
            "y": "1817.20",
            "width": "326.44",
            "height": "774.26",
            "confidence": None
        },
        {
            "id": "9",
            "label": "designation",
            "x": "529.59",
            "y": "1814.96",
            "width": "373.98",
            "height": "773.23",
            "confidence": None
        },
        {
            "id": "A",
            "label": "honoraire",
            "x": "804.52",
            "y": "1806.54",
            "width": "360.28",
            "height": "773.57",
            "confidence": None
        }
    ],
    "height": 2455,
    "key": "1430--9550775--20230720_page_0_1.jpg",
    "width": 1782
}
# new_temp
json_data = {
#     "boxes": [
#         {
#             "id": "1",
#             "label": "nom et prenom de adherent",
#             "x": "994.58",
#             "y": "280.82",
#             "width": "869.65",
#             "height": "103.09",
#             "confidence": None
#         },
#         {
#             "id": "3",
#             "label": "addresse de ladherent",
#             "x": "968.29",
#             "y": "424.44",
#             "width": "1042.39",
#             "height": "93.48",
#             "confidence": None
#         },
#         {
#             "id": "2",
#             "label": "numero cin ou passeport",
#             "x": "834.37",
#             "y": "353.79",
#             "width": "649.36",
#             "height": "85.95",
#             "confidence": None
#         },
#         {
#             "id": "4",
#             "label": "matricule cnam",
#             "x": "762.47",
#             "y": "487.58",
#             "width": "684.40",
#             "height": "87.42",
#             "confidence": None
#         },
#         {
#             "id": "5",
#             "label": "matricule de adherent",
#             "x": "976.32",
#             "y": "561.98",
#             "width": "1040.58",
#             "height": "92.86",
#             "confidence": None
#         },
#         {
#             "id": "6",
#             "label": "nom et prenom du malade",
#             "x": "972.04",
#             "y": "1175.40",
#             "width": "903.24",
#             "height": "125.34",
#             "confidence": None
#         },
#         {
#             "id": "7",
#             "label": "date de naissance",
#             "x": "947.17",
#             "y": "1249.44",
#             "width": "1107.46",
#             "height": "94.60",
#             "confidence": None
#         },
#         {
#             "id": "8",
#             "label": "date",
#             "x": "277.87",
#             "y": "1806.03",
#             "width": "409.10",
#             "height": "796.46",
#             "confidence": None
#         },
#         {
#             "id": "9",
#             "label": "designation",
#             "x": "508.23",
#             "y": "1764.35",
#             "width": "360.59",
#             "height": "716.67",
#             "confidence": None
#         },
#         {
#             "id": "A",
#             "label": "honoraire",
#             "x": "820.98",
#             "y": "1797.98",
#             "width": "348.12",
#             "height": "783.93",
#             "confidence": None
#         }
#     ],
#     "height": 2460,
#     "key": "0710--8574221--20230705_page_0.jpg",
#     "width": 3535
# }
# old_temp
# json_data = {
#     "boxes": [
#         {
#             "id": "1",
#             "label": "nom et prenom de adherent",
#             "x": "969.20",
#             "y": "322.10",
#             "width": "821.76",
#             "height": "107.85",
#             "confidence": None
#         },
#         {
#             "id": "2",
#             "label": "numero cin ou passeport",
#             "x": "824.60",
#             "y": "400.82",
#             "width": "595.76",
#             "height": "87.71",
#             "confidence": None
#         },
#         {
#             "id": "3",
#             "label": "addresse de ladherent",
#             "x": "951.59",
#             "y": "468.70",
#             "width": "1018.29",
#             "height": "106.09",
#             "confidence": None
#         },
#         {
#             "id": "4",
#             "label": "addresse de ladherent",
#             "x": "894.84",
#             "y": "546.97",
#             "width": "1560.32",
#             "height": "118.16",
#             "confidence": None
#         },
#         {
#             "id": "5",
#             "label": "matricule de adherent",
#             "x": "842.66",
#             "y": "626.58",
#             "width": "607.80",
#             "height": "101.27",
#             "confidence": None
#         },
#         {
#             "id": "6",
#             "label": "nom et prenom du malade",
#             "x": "1002.76",
#             "y": "1178.14",
#             "width": "771.49",
#             "height": "107.33",
#             "confidence": None
#         },
#         {
#             "id": "7",
#             "label": "date de naissance",
#             "x": "981.23",
#             "y": "1255.20",
#             "width": "1026.41",
#             "height": "104.94",
#             "confidence": None
#         },
#         {
#             "id": "8",
#             "label": "date",
#             "x": "266.78",
#             "y": "1639.05",
#             "width": "383.98",
#             "height": "459.52",
#             "confidence": None
#         },
#         {
#             "id": "9",
#             "label": "designation",
#             "x": "519.06",
#             "y": "1644.51",
#             "width": "324.29",
#             "height": "442.13",
#             "confidence": None
#         },
#         {
#             "id": "A",
#             "label": "honoraire",
#             "x": "817.59",
#             "y": "1651.59",
#             "width": "345.77",
#             "height": "460.33",
#             "confidence": None
#         },
#         {
#             "id": "B",
#             "label": "honoraire",
#             "x": "440.59",
#             "y": "2190.90",
#             "width": "618.41",
#             "height": "358.48",
#             "confidence": None
#         }
#     ],
#     "height": 2460,
#     "key": "0756--5637850--20230705_page_0.jpg",
#     "width": 3540
# }


SyntaxError: incomplete input (1811845983.py, line 410)

In [67]:

def convert_json_to_yolo(json_data, output_dir="labels"):

    class_mapping = {
        "nom et prenom de adherent": 0,
        "matricule cnam": 1,
        "matricule de adherent": 2,
        "addresse de ladherent": 3,
        "numero cin ou passeport": 4,
        "nom et prenom du malade": 5,
        "date de naissance": 6,
        "date": 7,
        "designation": 8,
        "honoraire": 9
    }

    # Extract image dimensions
    image_width = json_data["width"]
    image_height = json_data["height"]

    # Create the output directory if it doesn't exist
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    # Determine output file path from the key
    output_file = os.path.join(output_dir, os.path.splitext(json_data["key"])[0] + ".txt")
    
    # Write YOLO formatted annotations
    with open(output_file, "w") as f:
        for box in json_data["boxes"]:
            try:
                # Convert box coordinates from strings to float
                center_x = float(box["x"])
                center_y = float(box["y"])
                width = float(box["width"])
                height = float(box["height"])

                # Normalize
                center_x_norm = center_x / image_width
                center_y_norm = center_y / image_height
                width_norm = width / image_width
                height_norm = height / image_height

                # Get the class ID
                label = box["label"]
                if label not in class_mapping:
                    print(f"Warning: Label '{label}' not found in class mapping. Skipping.")
                    continue
                class_id = class_mapping[label]

                # Write to file
                f.write(f"{class_id} {center_x_norm:.6f} {center_y_norm:.6f} {width_norm:.6f} {height_norm:.6f}\n")
            except Exception as e:
                print(f"Error processing box with ID {box.get('id')}: {e}")

    print(f"Saved YOLO annotations to {output_file}")


In [68]:
source_folder = "data_to_annotate/new_temp1"
for filename in os.listdir(source_folder):
    if filename.lower().endswith((".jpg", ".jpeg", ".png")):
        path = os.path.join(source_folder, filename)
        json_data["key"]=filename
        convert_json_to_yolo(json_data)
        print(f'label created : {filename}')

Saved YOLO annotations to labels\0675--9124833--20230705_page_0_1.txt
label created : 0675--9124833--20230705_page_0_1.jpg
Saved YOLO annotations to labels\0706--8979887--20230705_page_0_1.txt
label created : 0706--8979887--20230705_page_0_1.jpg
Saved YOLO annotations to labels\0707--8979516--20230705_page_0_1.txt
label created : 0707--8979516--20230705_page_0_1.jpg
Saved YOLO annotations to labels\0708--8787562--20230705_page_0_1.txt
label created : 0708--8787562--20230705_page_0_1.jpg
Saved YOLO annotations to labels\0710--8574221--20230705_page_0_1.txt
label created : 0710--8574221--20230705_page_0_1.jpg
Saved YOLO annotations to labels\0711--8483891--20230705_page_0_1.txt
label created : 0711--8483891--20230705_page_0_1.jpg
Saved YOLO annotations to labels\0711--9459343--20230731_page_0_1.txt
label created : 0711--9459343--20230731_page_0_1.jpg
Saved YOLO annotations to labels\0712--9459344--20230731_page_0_1.txt
label created : 0712--9459344--20230731_page_0_1.jpg
Saved YOLO annot

In [61]:
# import os
# import cv2


# def annotate_image(image_path, label_path, output_dir="output_annotated"):
# # Define the class names globally or pass them as a parameter if needed
#     class_names = [
#         "nom et prenom de adherent",
#         "matricule cnam",
#         "matricule de adherent",
#         "addresse de ladherent",
#         "numero cin ou passeport",
#         "nom et prenom du malade",
#         "date de naissance",
#         "date",
#         "designation",
#         "honoraire"
#     ]

#     # Load the image
#     image = cv2.imread(image_path)
#     if image is None:
#         raise FileNotFoundError(f"Could not load image at {image_path}")

#     # Get image dimensions
#     height, width = image.shape[:2]

#     # Read the YOLO annotations
#     with open(label_path, "r") as f:
#         lines = f.readlines()

#     # Draw each bounding box on the image
#     for line in lines:
#         # Parse the YOLO annotation line
#         class_id, center_x, center_y, box_width, box_height = map(float, line.strip().split())
#         class_id = int(class_id)

#         # Convert normalized coordinates to pixel coordinates
#         x_min = int((center_x - box_width / 2) * width)
#         y_min = int((center_y - box_height / 2) * height)
#         x_max = int((center_x + box_width / 2) * width)
#         y_max = int((center_y + box_height / 2) * height)

#         # Draw the bounding box
#         color = (0, 255, 0)  # Green color in BGR
#         thickness = 2
#         cv2.rectangle(image, (x_min, y_min), (x_max, y_max), color, thickness)

#         # Add the class label above the box
#         label = class_names[class_id]
#         cv2.putText(image, label, (x_min, y_min - 10), cv2.FONT_HERSHEY_SIMPLEX, 0.5, color, thickness)

#     # Create output directory if it doesn't exist
#     if not os.path.exists(output_dir):
#         os.makedirs(output_dir)

#     # Generate the output path
#     file_name = os.path.basename(image_path)
#     name, ext = os.path.splitext(file_name)
#     output_path = os.path.join(output_dir, f"{name}_annotated{ext}")

#     # Save the annotated image
#     cv2.imwrite(output_path, image)
#     print(f"Saved annotated image to {output_path}")
#     return output_path


In [ ]:
# source_folder = "data_to_annotate/new_temp1"
# for filename in os.listdir(source_folder):
#     if filename.lower().endswith((".jpg", ".jpeg", ".png")):
#         path = os.path.join(source_folder, filename)
#         image_path = source_folder+"/"+filename
#         name_without_extension = os.path.splitext(filename)[0]
#         label_path = f"labels/{name_without_extension}.txt"
#         annotate_image(image_path, label_path)
#         print(f'Bounding box created : {filename}')


## Creation of dataset structure
##### data/
##### ├── dataset/
##### │   ├── images/
##### │   │   ├── train/
##### │   │   │   ├── image1.jpg
##### │   │   │   └── ...
##### │   │   ├── val/
##### │   │   │   ├── image2.jpg
##### │   │   │   └── ...
##### │   │   └── test/
##### │   │       ├── 0712--9550810--20230705_page_0.jpg
##### │   │       └── ...
##### │   ├── labels/
##### │   │   ├── train/
##### │   │   │   ├── image1.txt
##### │   │   │   └── ...
##### │   │   ├── val/
##### │   │   │   ├── image2.txt
##### │   │   │   └── ...
##### │   │   └── test/
##### │   │       ├── 0712--9550810--20230705_page_0.txt
##### │   │       └── ...
##### │   └── data.yaml

In [69]:
import os
import random
import shutil

# Define paths
image_dir = "Medical_form"
label_dir = "labels"
output_dir = "dataset"

# Define split ratios
train_ratio = 0.7  # 70% for training
val_ratio = 0.2    # 20% for validation
test_ratio = 0.1   # 10% for testing

# Define class names (assuming a single class for handwritten regions)
class_names = [
    "nom et prenom de adherent",
    "matricule cnam",
    "matricule de adherent",
    "addresse de ladherent",
    "numero cin ou passeport",
    "nom et prenom du malade",
    "date de naissance",
    "date",
    "designation",
    "honoraire"
]

# Create output directories
image_train_dir = os.path.join(output_dir, "images", "train")
image_val_dir = os.path.join(output_dir, "images", "val")
image_test_dir = os.path.join(output_dir, "images", "test")
label_train_dir = os.path.join(output_dir, "labels", "train")
label_val_dir = os.path.join(output_dir, "labels", "val")
label_test_dir = os.path.join(output_dir, "labels", "test")

# Create the directories if they don't exist
os.makedirs(image_train_dir, exist_ok=True)
os.makedirs(image_val_dir, exist_ok=True)
os.makedirs(image_test_dir, exist_ok=True)
os.makedirs(label_train_dir, exist_ok=True)
os.makedirs(label_val_dir, exist_ok=True)
os.makedirs(label_test_dir, exist_ok=True)

# Get list of all images
image_files = [f for f in os.listdir(image_dir) if f.endswith((".jpg", ".png", ".jpeg"))]
random.shuffle(image_files)  # Shuffle the list to randomize the split

# Calculate split indices
total_images = len(image_files)
train_end = int(total_images * train_ratio)
val_end = train_end + int(total_images * val_ratio)

# Split the images
train_files = image_files[:train_end]
val_files = image_files[train_end:val_end]
test_files = image_files[val_end:]

# Function to copy files to the appropriate directory
def copy_files(file_list, src_image_dir, src_label_dir, dst_image_dir, dst_label_dir):
    for file_name in file_list:
        # Copy image
        src_image_path = os.path.join(src_image_dir, file_name)
        dst_image_path = os.path.join(dst_image_dir, file_name)
        shutil.copy2(src_image_path, dst_image_path)
        
        # Copy corresponding label file
        label_file = os.path.splitext(file_name)[0] + ".txt"
        src_label_path = os.path.join(src_label_dir, label_file)
        dst_label_path = os.path.join(dst_label_dir, label_file)
        if os.path.exists(src_label_path):
            shutil.copy2(src_label_path, dst_label_path)
        else:
            print(f"Warning: Label file {src_label_path} not found for image {file_name}")

# Copy files to their respective directories
print("Copying training files...")
copy_files(train_files, image_dir, label_dir, image_train_dir, label_train_dir)
print("Copying validation files...")
copy_files(val_files, image_dir, label_dir, image_val_dir, label_val_dir)
print("Copying test files...")
copy_files(test_files, image_dir, label_dir, image_test_dir, label_test_dir)

# Create data.yaml file
data_yaml = f"""
train: {os.path.abspath(image_train_dir)}
val: {os.path.abspath(image_val_dir)}
test: {os.path.abspath(image_test_dir)}

nc: {len(class_names)}
names: {class_names}
"""

# Save data.yaml
data_yaml_path = os.path.join(output_dir, "data.yaml")
with open(data_yaml_path, "w") as f:
    f.write(data_yaml)

print(f"Dataset split complete. Created {data_yaml_path}")
print(f"Train: {len(train_files)} images")
print(f"Validation: {len(val_files)} images")
print(f"Test: {len(test_files)} images")

Copying training files...
Copying validation files...
Copying test files...
Dataset split complete. Created dataset\data.yaml
Train: 910 images
Validation: 260 images
Test: 131 images


## installing libraries for YOLOv8

In [70]:
# !pip install ultralytics

In [71]:
!yolo --version

WARNING ⚠� argument '--version' does not require leading dashes '--', updating to 'version'.
8.3.107


## fine tunning model

In [72]:
from ultralytics import YOLO

# Load the pretrained YOLOv8 model
model = YOLO("yolov8s.pt")  # Small model for better accuracy

# Fine-tune the model on your dataset
model.train(
    data="dataset/data.yaml",  # Path to your dataset
    epochs=10,                # Number of epochs
    imgsz=640,                 # Image size (try 1280 if GPU allows)
    batch=16,                  # Batch size (reduce to 8 if memory issues)
    device=None,                  # Use GPU
    patience=5,               # Early stopping patience
    optimizer="AdamW",         # Optimizer for fine-tuning
    lr0=0.001,                 # Initial learning rate (lower for fine-tuning)
    cos_lr=True,               # Cosine learning rate scheduling (t7otha tsaghar learning rate)
    weight_decay=0.0005,       # Weight decay for regularization (valeur li toz8er biha)
    
    # Data augmentation
    hsv_h=0.015,               # Hue augmentation (colors)
    hsv_s=0.7,                 # Saturation augmentation (color intensity vibrant)
    hsv_v=0.4,                 # Brightness augmentation
    degrees=0.0,              # Random rotation
    translate=0.1,             # Random translation
    scale=0.5,                 # Random scaling
    shear=0.0,                 # Random shearing
    mosaic=0.0,                # Mosaic augmentation combine 4 images into 1 
    
    # Model-specific
    freeze=10,                 # Freeze first 10 layers (backbone)
    pretrained=True,           # Use pretrained weights
    
    # Logging
    name="medical_form_yolo_one_temp",  # Name of the training run
    save_period=-1             # Save only best and last weights
)

print("Fine-tuning complete. Model weights saved in runs/detect/medical_form_yolo_one_temp/weights/")

New https://pypi.org/project/ultralytics/8.3.108 available  Update with 'pip install -U ultralytics'
Ultralytics 8.3.107  Python-3.11.11 torch-2.3.1 CPU (13th Gen Intel Core(TM) i7-13620H)
engine\trainer: task=detect, mode=train, model=yolov8s.pt, data=dataset/data.yaml, epochs=10, time=None, patience=5, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=medical_form_yolo_one_temp, exist_ok=False, pretrained=True, optimizer=AdamW, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=True, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=10, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False

train: Scanning C:\Users\User\OneDrive - ESPRIT\Bureau\4DS\Semester2\Project\ProjectDS\modelling\dataset\labels\train... 910 images, 0 backgrounds, 0 corrupt: 100%|██████████| 910/910 [00:00<00:00, 1156.29it/s]


train: New cache created: C:\Users\User\OneDrive - ESPRIT\Bureau\4DS\Semester2\Project\ProjectDS\modelling\dataset\labels\train.cache


val: Scanning C:\Users\User\OneDrive - ESPRIT\Bureau\4DS\Semester2\Project\ProjectDS\modelling\dataset\labels\val... 260 images, 0 backgrounds, 0 corrupt: 100%|██████████| 260/260 [00:00<00:00, 1192.68it/s]

val: New cache created: C:\Users\User\OneDrive - ESPRIT\Bureau\4DS\Semester2\Project\ProjectDS\modelling\dataset\labels\val.cache
Plotting labels to runs\detect\medical_form_yolo_one_temp\labels.jpg... 


optimizer: AdamW(lr=0.001, momentum=0.937) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 0 dataloader workers
Logging results to runs\detect\medical_form_yolo_one_temp
Starting training for 10 epochs...
Closing dataloader mosaic

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       1/10         0G       2.03      2.515      1.777        139        640: 100%|██████████| 57/57 [08:01<00:00,  8.45s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 9/9 [01:09<00:00,  7.70s/it]

                   all        260       2600      0.463      0.507      0.512      0.233



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       2/10         0G      1.539      1.427      1.436        139        640: 100%|██████████| 57/57 [07:39<00:00,  8.07s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 9/9 [01:09<00:00,  7.75s/it]

                   all        260       2600      0.639      0.578      0.646      0.337



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       3/10         0G      1.369      1.115      1.297        134        640: 100%|██████████| 57/57 [08:07<00:00,  8.55s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 9/9 [01:10<00:00,  7.80s/it]

                   all        260       2600      0.755        0.7      0.802      0.433



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       4/10         0G      1.237     0.9351      1.212        139        640: 100%|██████████| 57/57 [07:42<00:00,  8.12s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 9/9 [01:03<00:00,  7.03s/it]

                   all        260       2600       0.79      0.745      0.847      0.535



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       5/10         0G       1.16     0.8484      1.166        135        640: 100%|██████████| 57/57 [07:55<00:00,  8.35s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 9/9 [01:09<00:00,  7.67s/it]

                   all        260       2600      0.662      0.722      0.765      0.477



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       6/10         0G       1.11     0.7933      1.127        138        640: 100%|██████████| 57/57 [07:38<00:00,  8.04s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 9/9 [01:04<00:00,  7.15s/it]

                   all        260       2600        0.7      0.788      0.812      0.541



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       7/10         0G      1.053     0.7211      1.091        140        640: 100%|██████████| 57/57 [07:52<00:00,  8.28s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 9/9 [01:10<00:00,  7.80s/it]

                   all        260       2600      0.727      0.812       0.82      0.544



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       8/10         0G     0.9938     0.6683       1.07        138        640: 100%|██████████| 57/57 [07:55<00:00,  8.34s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 9/9 [01:12<00:00,  8.01s/it]

                   all        260       2600      0.821      0.869      0.899      0.618



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       9/10         0G     0.9586     0.6352      1.051        134        640: 100%|██████████| 57/57 [07:59<00:00,  8.41s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 9/9 [01:10<00:00,  7.83s/it]

                   all        260       2600      0.687       0.79      0.808      0.558



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      10/10         0G      0.932     0.6202      1.039        137        640: 100%|██████████| 57/57 [07:36<00:00,  8.01s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 9/9 [01:10<00:00,  7.79s/it]

                   all        260       2600      0.755       0.85       0.87      0.603



10 epochs completed in 1.501 hours.
Optimizer stripped from runs\detect\medical_form_yolo_one_temp\weights\last.pt, 22.5MB
Optimizer stripped from runs\detect\medical_form_yolo_one_temp\weights\best.pt, 22.5MB

Validating runs\detect\medical_form_yolo_one_temp\weights\best.pt...
Ultralytics 8.3.107  Python-3.11.11 torch-2.3.1 CPU (13th Gen Intel Core(TM) i7-13620H)
Model summary (fused): 72 layers, 11,129,454 parameters, 0 gradients, 28.5 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 9/9 [01:07<00:00,  7.48s/it]


                   all        260       2600      0.821      0.868      0.899      0.618
nom et prenom de adherent        260        260       0.81      0.838      0.888      0.572
        matricule cnam        260        260       0.88        0.8      0.886      0.508
 matricule de adherent        260        260      0.835      0.917      0.921       0.61
 addresse de ladherent        260        260      0.808      0.704      0.825       0.53
numero cin ou passeport        260        260      0.809      0.798      0.845      0.504
nom et prenom du malade        260        260      0.639      0.869      0.857      0.565
     date de naissance        260        260      0.635      0.855      0.817      0.498
                  date        260        260      0.963      0.998      0.995      0.872
           designation        260        260      0.917      0.962      0.986      0.772
             honoraire        260        260      0.917       0.94      0.968      0.745
Speed: 1.6ms pre

In [51]:
import cv2
import os
from ultralytics import YOLO

# Load the trained YOLOv8 model
model_path = "runs/detect/medical_form_yolo_best2/weights/best.pt"
model = YOLO(model_path)

# Define class names (must match the order in data.yaml)
# Define class names (assuming a single class for handwritten regions)
class_names = [
    "nom et prenom de adherent",
    "matricule cnam",
    "matricule de adherent",
    "addresse de ladherent",
    "numero cin ou passeport",
    "nom et prenom du malade",
    "date de naissance",
    "date",
    "designation",
    "honoraire"
]

# Create output folder for annotated images and cropped regions
output_folder = "predictions"
if not os.path.exists(output_folder):
    os.makedirs(output_folder)

# Path to the new image for prediction
image_path = "dataset/images/val/0756--5637850--20230705_page_0_1.jpg"  # Replace with the path to your new image
image = cv2.imread(image_path)
if image is None:
    raise FileNotFoundError(f"Could not load image at {image_path}")

# Run inference
results = model(image, imgsz=640, conf=0.5, device=None)  # Use GPU (device=0); set to None for CPU

# Process the detected bounding boxes
crop_counter = 1
for result in results:
    boxes = result.boxes.xyxy  # Get bounding boxes in [x_min, y_min, x_max, y_max] format
    confidences = result.boxes.conf  # Get confidence scores
    class_ids = result.boxes.cls  # Get class IDs

    for box, conf, cls_id in zip(boxes, confidences, class_ids):
        # Extract bounding box coordinates
        x1, y1, x2, y2 = map(int, box)
        class_id = int(cls_id)
        label = class_names[class_id]
        confidence = float(conf)

        # Draw the bounding box on the image
        color = (0, 255, 0)  # Green color in BGR
        thickness = 2
        cv2.rectangle(image, (x1, y1), (x2, y2), color, thickness)
        cv2.putText(image, f"{label} ({confidence:.2f})", (x1, y1 - 10), cv2.FONT_HERSHEY_SIMPLEX, 0.5, color, thickness)

        # Apply 20px vertical padding for cropping
        height, width = image.shape[:2]
        padding = 20
        x1 = max(0, x1)
        x2 = min(width, x2)
        y1 = max(0, y1 - padding)
        y2 = min(height, y2 + padding)

        # Crop the detected region
        if x2 > x1 and y2 > y1:
            cropped_image = image[y1:y2, x1:x2]
            
            # Save the cropped image
            output_filename = f"{label.replace(' ', '_')}_{crop_counter}.jpg"
            output_path = os.path.join(output_folder, output_filename)
            cv2.imwrite(output_path, cropped_image)
            print(f"Saved cropped region to {output_path}")
            crop_counter += 1

# Save the annotated image
image_filename = os.path.basename(image_path)
annotated_image_path = os.path.join(output_folder, f"annotated_{image_filename}")
cv2.imwrite(annotated_image_path, image)
print(f"Saved annotated image to {annotated_image_path}")




0: 640x480 1 nom et prenom de adherent, 1 addresse de ladherent, 1 nom et prenom du malade, 1 date de naissance, 1 date, 2 honoraires, 140.8ms
Speed: 2.9ms preprocess, 140.8ms inference, 0.9ms postprocess per image at shape (1, 3, 640, 480)
Saved cropped region to predictions\date_1.jpg
Saved cropped region to predictions\date_de_naissance_2.jpg
Saved cropped region to predictions\honoraire_3.jpg
Saved cropped region to predictions\nom_et_prenom_de_adherent_4.jpg
Saved cropped region to predictions\addresse_de_ladherent_5.jpg
Saved cropped region to predictions\honoraire_6.jpg
Saved cropped region to predictions\nom_et_prenom_du_malade_7.jpg
Saved annotated image to predictions\annotated_0756--5637850--20230705_page_0_1.jpg
